core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -57,6 +59,137 @@ object Scheduler {
+
+    // We need to handle the case when the source is canceled or failed
+    stateQueue.watchCompletion().onComplete {
+      case Success(_) =>
+        killSwitch.shutdown()
+      case Failure(cause) =>
+        killSwitch.abort(cause)
+    }
+
+    def sourceFromSinkQueue[T](queue: SinkQueueWithCancel[T]): Source[T, NotUsed] = {
+      Source
+        .unfoldResourceAsync[T, SinkQueueWithCancel[T]](
+          create = () => Future.successful(queue),
+          read = queue => queue.pull(),
+          close = queue =>
+            Future.successful {
+              killSwitch.shutdown()
+              queue.cancel()
+              Done
+          })
+    }
+
+    Source.maybe.prepend {
+      val events = sourceFromSinkQueue(specQueue)
+        .watchTermination() {
+          case (_, completionSignal) =>
+            completionSignal.onComplete {
+              case Success(_) =>
+                killSwitch.shutdown()
+              case Failure(cause) =>
+                killSwitch.abort(cause)
+            }
+        }
+
+      Source.single(specsSnapshot -> events)
+    }.via(schedulerFlow)
+      .flatMapConcat {
+        case (snapshot, updates) =>
+          stateSnapshotPromise.trySuccess(snapshot)
+          updates.watchTermination() {
+            case (_, cancellationSignal) =>
+              cancellationSignal.onComplete {
+                case Success(_) =>
+                  killSwitch.shutdown()
+                case Failure(cause) =>
+                  killSwitch.abort(cause)
+              }
+          }
+      }
+      .mapAsync(1)(stateQueue.offer)
+      .map {
+        case QueueOfferResult.Enqueued =>
+        case QueueOfferResult.QueueClosed =>
+          killSwitch.shutdown()
+        case QueueOfferResult.Failure(cause) =>
+          killSwitch.abort(cause)
+        case QueueOfferResult.Dropped => // we shouldn't receive that at all because OverflowStrategy.backpressure
+          throw new RuntimeException("Unexpected QueueOfferResult.Dropped element")
+      }
+      .runWith(Sink.ignore)
+
+    val sourceWithKillSwitch = stateSource
+      .watchTermination() {
+        case (materializedValue, cancellationSignal) =>
+          cancellationSignal.onComplete {
+            case Success(_) =>
+              killSwitch.shutdown()
+            case Failure(cause) =>
+              killSwitch.abort(cause)
+          }
+          materializedValue
+      }
+      .via(killSwitch.flow)
+
+    val sinkWithKillSwitch = Flow[SpecUpdated]
+      .watchTermination() {
+        case (materializedValue, cancellationSignal) =>
+          cancellationSignal.onComplete {
+            case Success(_) =>
+              killSwitch.shutdown()
+            case Failure(cause) =>
+              killSwitch.abort(cause)
+          }
+          materializedValue
+      }
+      .via(killSwitch.flow)
+      .to(specSink)
+
+    (stateSnapshotPromise.future, sourceWithKillSwitch, sinkWithKillSwitch)
```

Comment:
Could you add a few comments in between. It is not easy to grasp the whole thing.